### PR TITLE
Minimizing the rebuild of sticky scroll

### LINKED ISSUE #192535: Sticky scroll: folding icons missing from some sticky headers
Type: <b>Bug</b>

> Issue troubleshooting has identified that the issue is with Visual Studio Code - Insiders. 

If the sticky scroll header shows only one line, that line has no folding indicator.

If it shows 2 or more lines, only the last line has the indicator.

junk

This is working correctly in 1.82.0 Stable.

VS Code version: Code - Insiders 1.83.0-insider (5a400e53e985dc5e24f6ee574b07ab23943841c5, 2023-09-07T17:35:42.740Z)
OS version: Windows_NT x64 10.0.22621
Modes:

### PR DESCRIPTION
**Pull request by aiday-mar**
Part of the fix for https://github.com/microsoft/vscode/issues/191319

Fixes: https://github.com/microsoft/vscode/issues/192535. This issue has appeared recently, the fix is to set : `display :inline-block` on the inner line number dom node. This change also made it unecessary to have a CSS padding right on the folding toggle icon which was removed.

### CODE REVIEW

File: src/vs/editor/contrib/stickyScroll/browser/stickyScrollWidget.ts [hunk 1]
```diff
@@ -109,7 +110,8 @@ export class StickyScrollWidget extends Disposable implements IOverlayWidget {
 		return this._lineNumbers;
 	}
 
-	setState(state: StickyScrollWidgetState | undefined): void {
+	setState(state: StickyScrollWidgetState | undefined, forceRebuild: boolean = false): void {
+		this._previousStickyLines = [...this._stickyLines];
```
Comment (aeschli) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScrollWidget.ts:114]: By storing it in a field, this array will be referenced also after `setState``, increasing the referenced memory. So I would only do that if needed. Here it's probably enough to pass it as an argument to `_renderRootNode`.
Also you create a copy of the array. That's not needed as `_clearStickyWidget` will just create a new array.

**Comment by aeschli** [comment]
Folding is currently broken, it works only once. 
On click, sticky scroll is rebuilt and all click listeners are desposed, even if the line can be reused.

I was wondering if it's a better idea to have a single on click listener, on the whole sticky scroll composite. Then find out of it's over a folding indicator and modify the corresponding widget.

Also I was thinking that the code for the folding icon transition should get simpler and no trick is needed to avoid the flickering.

**Comment by aiday-mar** [comment]
Thank you for your valuable feedback, there are some good points. I changed the code so as to reuse the single on click listener from the controller. It checks if the attribute `data-sticky-line-is-folding-icon` is set on the mouseEvent.target and if yes, it will toggle the folding region. 

The reason why the code still does a check of whether it is on a folding icon, is because I forgot to merge the main branch. The PR https://github.com/microsoft/vscode/pull/191440 merged a couple of days ago, aimed to simplify precisely the code for the folding icon transitions.

**Comment by aeschli** [comment]
When I click on an expanded folding icon, the code is collapsed, but the the icon is still uncollapsed.
Also, if there's a line in the sticky scroll that does not have a folding icon, clicking on it does something.
Here's my code sample: Scroll inside the `div`.
```
<html>
<head>
	<meta charset="utf-8">
	<title>VSCode Tests</title>
	<link href="https://cdn.rawgit.com/mochajs/mocha/2.2.5/mocha.css" rel="stylesheet" />

	<style type="text/css">
		body {
			color: purple;
			background-color: #d8da3d;
		}
	</style>
</head>
<body>
	<div id="mocha"></div>
	<blockquote cite="fooo.sdsd">


	<script>
		mocha.setup('tdd');

		require.config({
			baseUrl: '/out',
			paths: {
				assert: '/test/assert.js'
			}
		});

		require({{ modules }}, function () {
			mocha.run();
		});
	</script>
	<div class="js-stale-session-flash stale-session-flash flash flash-warn flash-banner hidden">
		<span class=codicon></span>
		<span class="signed-in-tab-flash">You signed in with another tab or window. <a href="">Reload</a> to refresh your session.</span>
		<span class="signed-out-tab-flash">You signed out in another tab or window. <a href="">Reload</a> to refresh your session.</span>
	</div>
</body>
</html>
```

**Comment by aiday-mar** [comment]
HI thanks for the example, I tested on the example you provided and was able to reproduce the error. For some reason while testing on my local typescript example files, the folding appears to work fine. I'll investigate.

**Review by aeschli — DISMISSED** [review]
Looks good!

**Comment by aiday-mar** [comment]
PR has merge conflict at the point where merge was to be done. The merge conflict has been resolved.

**Review by aeschli — APPROVED** [review]

**Pull request merged**